### ISSUE TITLE
Can't read-your-writes over isolate if cached query stream has existing listener

### ISSUE THREAD

**Issue by Mike278** [simolus3/moor#2166]
We're finally slowwwly migrating from `package:sqflite` to `package:sqlite3` with isolates, and we came across some interesting test failures.

Normally you're able to read your writes - i.e. after writing to the database you're able to immediately read the results of that write. However this doesn't seem to be true for the specific case where you're using an isolate and the "read" query is a cache hit on a stream with an existing listener. 

This is easier to show with a test: 

7/8 tests pass, with the failing test being `existingListener=true, useIsolate=true, useTransaction=false`. Note that 8/8 tests pass if you allow some time for the update event to propagate by uncommenting `// await pumpEventQueue();`
```dart
import 'dart:isolate';

import 'package:drift/isolate.dart';
import 'package:drift/native.dart';
import 'package:drift/drift.dart';
import 'package:test/test.dart';

part 'main.g.dart';

void main() {
  driftRuntimeOptions.dontWarnAboutMultipleDatabases = true;
  for (final existingListener in [true, false]) {
    for (final useIsolate in [true, false]) {
      for (final useTransaction in [true, false]) {
        final vars = 'existingListener=$existingListener, '
            'useIsolate=$useIsolate, '
            'useTransaction=$useTransaction';
        
        test('can read-your-writes ($vars)', () async {
          final db = useIsolate
              ? _SomeDb.connect(await _connect())
              : _SomeDb(NativeDatabase.memory(logStatements: true));

          await db.into(db.someTable).insert(_SomeTableDataCompanion());

          Stream<_SomeTableData> getRow() =>
              db.select(db.someTable).watchSingle();

          Future<void> readYourWrite() async {
            final update = _SomeTableDataCompanion(name: Value('x'));
            await db.update(db.someTable).write(update);
            // await pumpEventQueue();
            final row = await getRow().first;
            expect(row.name, equals('x'),
                reason: 'should be able to read the row we just wrote');
          }

          if (existingListener) {
            getRow().listen(null);
          }

          await (useTransaction
              ? db.transaction(readYourWrite)
              : readYourWrite());
        });
      }
    }
  }
}

Future<DatabaseConnection> _connect() async {
  final out = ReceivePort();
  final args = out.sendPort;
  await Isolate.spawn(_isolateEntrypoint, args);
  final isolate = (await out.first) as DriftIsolate;
  return isolate.connect(isolateDebugLog: true);
}

void _isolateEntrypoint(SendPort out) {
  final driftIsolate = DriftIsolate.inCurrent(() {
    final driver = NativeDatabase.memory(logStatements: true);
    return DatabaseConnection.fromExecutor(driver);
  });
  out.send(driftIsolate);
}

class _SomeTable extends Table {
  IntColumn get id => integer().autoIncrement()();

  TextColumn get name => text().nullable()();
}

@DriftDatabase(tables: [_SomeTable])
class _SomeDb extends _$_SomeDb {
  _SomeDb(super.executor);

  _SomeDb.connect(DatabaseConnection connection) : super.connect(connection);
  @override
  final schemaVersion = 1;
}
```

Initially this seemed like a bug because it works in the other 7 cases, but the more I think about it as I type this out I'm not so sure. 

Consistency across multiple statements outside of a transaction can't be guaranteed, and maybe the delay caused by the isolate + delivering an event to the existing stream listener is enough to expose that. 

On the other hand there's only one sqlite connection. If there was a second read-only connection I might expect to read stale results for a short amount of time after an update, but not with a single connection. Maybe there's some race condition with how query streams invalidate their cached results over isolate?

In any case, this is only an issue in test code. In real app code anything that depends on consistency across multiple statements is always wrapped in a transaction.

**Comment by simolus3** [comment]
TL;DR: Isolate communication is a distributed system and distributed systems are hard.

Due to scheduling, the test can run in slightly different ways. But essentially, the problem boils down to this trace:

1. We run the insert and also send a notification indicating that there was an insert on the table.
2. The `await insert` completes before the notification comes back, so the `if (existingListener)` part subscribes and runs the `SELECT *` (query 1)
3. We send the update to the remote isolate.
4. In the meantime, but before the update completes, we get back the insert notification from step 1. The active listener will cause another select to be issued (query 2).
5. We now get the select results for query 1. These results are forwarded to active listeners, __which is a bug__. They should be dropped since query 1 got stale by the time it completed. This result (`{id: 1, name: null}`) is also stored as the currently-active snapshot for that query.
6. The server informs us that the update has been successful. We send out the notification that the table was updated. However, we only invalidate queries after receiving this notification back from the remote isolate (which will broadcast it to all clients). So the local query is not invalidated. __This is also a bug.__
7. We run the `getRow().first`, which attaches a new listener to the query. There's an active snapshot (from step 5), which is immediately sent to this new listener, completing the `.first` future with the wrong value.

I think there are a number of improvements to be made here:

1. The first bug needs to be fixed (1d02ad9cd9d22a3bc3b0e6a16a2e22b4d2a48c24). We shouldn't consider results for queries that we issued before a write. I am a bit worried about additional load if _lots_ of frequent updates are happening. In that case, a stream query might be stuck in an invalidate-refresh cycle where no select ever gets completed without being invalidated first.  I think that's still better than emitting stale results like we're doing now, but it's a potential issue to be aware of. I'm not sure if there's an approach to solve this. I'll have to see how much this ends up being a problem in practice.
2. I have fixed the second bug in 9b0292f543dfad01f5a0de7887af5364b2ba1dcb. The table update message is forwarded to _other_ clients, there's no need for that indirection locally.
3. Drift 2.3.0 will introduce the concept of a "single client mode" to drift isolates (097f941f88ed95340eda32d9d44661d15029399c). In the common case where a database will only have one client isolate, we can avoid sending update notifications at all.

**Comment by Mike278** [comment]
Dang nice find! Distributed system + cache invalidation, always fun :)
